### PR TITLE
[MM-45991] Check and return JSON errors

### PR DESCRIPTION
**Pull request by noxer**
#### Summary
Throughout the server serialization errors are ignored and not returned or logged. This PR attaches most of them to a returned error or -- if that's not possible -- logs them.

#### Ticket Link
[MM-45991](https://mattermost.atlassian.net/browse/MM-45991)

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by agnivade** [comment]
I see a lot of code repeated from your last PR. Maybe a git issue?

**Comment by noxer** [comment]
@agnivade I was too enthusiastic and opened the PR too soon.

File: model/preference.go
Comment (agnivade) [inline model/preference.go:109]: Same here and other places as well.

File: model/utils.go
Comment (agnivade) [inline model/utils.go:341]: Let's leave it for now. I'd really want to avoid logging in the `model` package. One thing I noticed is that this function is mainly used in tests, and the places it's used in non-test code are handful. We can make an internal helper which returns an error and avoid calling this, and let the tests call this. 0/5, up to you.

File: model/config.go
Comment (noxer) [reply model/config.go:3187]: Hmm, in this case we might even get away with returning the error, right? Calling `json.NewDecoder` on the call side would require a similar change.

Comment (agnivade) [reply model/config.go:3187]: That will change the API. Changing any public API in the model package is a breaking change and we can't do that. This is root of all the problems :)

Comment (noxer) [reply model/config.go:3187]: Ah, okay. Yeah, then we just replace it at the caller's side.

File: model/utils.go
Comment (noxer) [reply model/utils.go:341]: If it's just for tests, let's drop the error.

Comment (agnivade) [reply model/utils.go:341]: Go for it :+1:

File: model/config.go
Comment (isacikgoz) [reply model/config.go:3187]: Alternatively, you can add a new method. In this case it's not so necessary though.

**Review by isacikgoz — COMMENTED** [review]
Good stuff, and nothing much to add on top of @agnivade's thorough review :)

**Comment by agnivade** [comment]
@noxer - Please feel free to request a review when this is ready. Thanks :)

**Comment by noxer** [comment]
Linter is currently complaining about focalboard...

**Comment by agnivade** [comment]
Oh dear, you've spent so much time changing all the variable names as well .. hopefully it wasn't too laborious. But great work anyways :)

**Comment by noxer** [comment]
Thanks. No worries, I went into Zen mode and did it mechanically.

File: api4/cloud.go
Comment (agnivade) [inline api4/cloud.go:128]: Just a thought. Since we are moving to the wrapping approach, we might want to create a custom "go vet" check that flags any new code added with the `err.Error()` field passed.

And then with the next breaking change, we can change the function signature to pass the error itself.

File: api4/config.go
Comment (agnivade) [inline api4/config.go:114]: What will happen if we have the `cfg == nil` case and pass a nil err? Nothing will panic right?

Same applies for the below cases as well.

**Review by agnivade — COMMENTED** [review]
This is a HUGE PR. I did a first pass. WIll do the next pass on Monday. Lots of good refactoring here. Great work!

File: api4/cloud.go
Comment (noxer) [reply api4/cloud.go:128]: Love the idea!

File: api4/config.go
Comment (noxer) [reply api4/config.go:114]: Nothing happens. It will set the wrapped error to `nil`, just as if no error was wrapped.

File: app/integration_action.go
Comment (agnivade) [inline app/integration_action.go:597]: This is actually being called from api4/integration action. So we might want to pass a request.CTX and call the logger instance of that. But out of scope for this PR of course.

File: app/post.go
Comment (agnivade) [inline app/post.go:254]: `c.Logger().Warn`

File: app/slashcommands/command_loadtest.go
Comment (agnivade) [inline app/slashcommands/command_loadtest.go:573]: Nice catch!

File: jobs/migrations/advanced_permissions_phase_2.go
Comment (agnivade) [inline jobs/migrations/advanced_permissions_phase_2.go:33]: Same comment as earlier. Let's not change this and `Decode` it at the call site.

File: model/preference.go
Comment (agnivade) [inline model/preference.go:108]: This call to `regexp.MustCompile` everytime is bad for performance. Can we move it outside and make it a global var? (Slightly out of scope but requires very minimal changes)

File: app/integration_action.go
Comment (noxer) [reply app/integration_action.go:597]: I agree with both. I'll keep this in and tackle it in another PR.

File: app/post.go
Comment (noxer) [reply app/post.go:254]: Nice catch. I missed the context argument.

**Review by agnivade — APPROVED** [review]

File: app/group.go
Comment (isacikgoz) [inline app/group.go:171]: Nice one 💯

**Review by isacikgoz — APPROVED** [review]
Nicely done!

**Comment by Phrynobatrachus** [comment]
Server fails to start with:

`Error relocating /mattermost/bin/mattermost: fcntl64: symbol not found`

I asked the release team about it since it seems like a build issue.

**Comment by phoinixgrr** [comment]
> Server fails to start with:
> 
> `Error relocating /mattermost/bin/mattermost: fcntl64: symbol not found`
> 
> I asked the release team about it since it seems like a build issue.

We performed an investigation on this. 
So it seems that the tests fail with the container image named: `mm-ee-test` failing to launch with: 
```
Digest: sha256:009c5fa21d9b40a71a092754a61bf35485a8952c55ac4f39988100eb09b0c2c6
Status: Downloaded newer image for mattermost/mm-ee-test:64b60f3
Error relocating /mattermost/bin/mattermost: fcntl64: symbol not found
```

Searching the web about the issue thrown, returns non deterministic answers to the issue presented. Some articles are proposing installing glibc when building using alpine images. (alpine uses musl libc) 

Found that the specific image commit [64b60f3](https://github.com/mattermost/mattermost-server/pull/20784/commits/64b60f31f1f34048a2dbb7e2a385d1896b44d1db) is introduced with PR: https://github.com/mattermost/mattermost-server/pull/20784

And the build & push of `mattermost/mm-ee-test:64b60f3` has being performed with  the following CircleCI pipeline https://app.circleci.com/pipelines/github/mattermost/mattermost-server/36036/workflows/5a4310ef-1956-4ea2-887e-a98f73d55e78/jobs/303271


Running the image produced by the above pipeline indeed I am able to reproduce the issue: 
```
docker run -it --platform linux/amd64 --rm mattermost/mm-ee-test:64b60f3
Error relocating /mattermost/bin/mattermost: fcntl64: symbol not found
```

Taking a look at the [latest CircleCI](https://app.circleci.com/pipelines/github/mattermost/mattermost-server?status=none&status=success) builds for `mattermost-server`, 

Seems like the following commit hashes: 

63a4545 
3a2d0ca 
f2658c2
f0c3eb5 

All belonging to PR [chore/cld-8000-add-focalboard-to-enterprise-repo](https://github.com/mattermost/mattermost-server/pull/20772) , maybe has introduced the issue. 
I am able to reproduce the issue on all commits after this PR merged, and none before. But this is a long shot on the actual culprint. It is difficult to actual locate the source of the issue, since we do not test the image produced with every build of the container image: `mm-ee-test` . 

As a first step, I would suggest reverting changes performed with `chore/cld-8000-add-focalboard-to-enterprise-repo` and see if that fixes the issue. 
@pfltdv  Would you mind confirming/dyning the above assumption, by reverting the change, and providing a build of `mm-ee-test` image? 

Thanks

**Comment by pfltdv** [comment]
> > Server fails to start with:
> > `Error relocating /mattermost/bin/mattermost: fcntl64: symbol not found`
> > I asked the release team about it since it seems like a build issue.
> 
> We performed an investigation on this. So it seems that the tests fail with the container image named: `mm-ee-test` failing to launch with:
> 
> ```
> Digest: sha256:009c5fa21d9b40a71a092754a61bf35485a8952c55ac4f39988100eb09b0c2c6
> Status: Downloaded newer image for mattermost/mm-ee-test:64b60f3
> Error relocating /mattermost/bin/mattermost: fcntl64: symbol not found
> ```
> 
> Searching the web about the issue thrown, returns non deterministic answers to the issue presented. Some articles are proposing installing glibc when building using alpine images. (alpine uses musl libc)
> 
> Found that the specific image commit [64b60f3](https://github.com/mattermost/mattermost-server/pull/20784/commits/64b60f31f1f34048a2dbb7e2a385d1896b44d1db) is introduced with PR: #20784
> 
> And the build & push of `mattermost/mm-ee-test:64b60f3` has being performed with the following CircleCI pipeline https://app.circleci.com/pipelines/github/mattermost/mattermost-server/36036/workflows/5a4310ef-1956-4ea2-887e-a98f73d55e78/jobs/303271
> 
> Running the image produced by the above pipeline indeed I am able to reproduce the issue:
> 
> ```
> docker run -it --platform linux/amd64 --rm mattermost/mm-ee-test:64b60f3
> Error relocating /mattermost/bin/mattermost: fcntl64: symbol not found
> ```
> 
> Taking a look at the [latest CircleCI](https://app.circleci.com/pipelines/github/mattermost/mattermost-server?status=none&status=success) builds for `mattermost-server`,
> 
> Seems like the following commit hashes:
> 
> [63a4545](https://github.com/mattermost/mattermost-server/commit/63a45458af683565e80e3ca4b4aba41c0a17e4e2) [3a2d0ca](https://github.com/mattermost/mattermost-server/commit/3a2d0cab9590765180a871bec99e5b4bea18756b) [f2658c2](https://github.com/mattermost/mattermost-server/commit/f2658c21934bf4d254060179c5e89eaca7cda707) [f0c3eb5](https://github.com/mattermost/mattermost-server/commit/f0c3eb5de684d8867cac9513243476cd42891c38)
> 
> All belonging to PR [chore/cld-8000-add-focalboard-to-enterprise-repo](https://github.com/mattermost/mattermost-server/pull/20772) , maybe has introduced the issue. I am able to reproduce the issue on all commits after this PR merged, and none before. But this is a long shot on the actual culprint. It is difficult to actual locate the source of the issue, since we do not test the image produced with every build of the container image: `mm-ee-test` .
> 
> As a first step, I would suggest reverting changes performed with `chore/cld-8000-add-focalboard-to-enterprise-repo` and see if that fixes the issue. @pfltdv Would you mind confirming/dyning the above assumption, by reverting the change, and providing a build of `mm-ee-test` image?
> 
> Thanks

3/5 that is related, since that PR only excludes board tests from server tests. But will test.

**Comment by pfltdv** [comment]
Tested with master tag and we have no similar issue with that image. (Published 18 minutes ago) And we do not have that issue for 8660461 (Published 41 minutes ago). Looks like we have one case trigger that failure. Still investigating.


```
mmm:~> docker run -it --platform linux/amd64 --rm mattermost/mm-ee-test:master
Unable to find image 'mattermost/mm-ee-test:master' locally
master: Pulling from mattermost/mm-ee-test
8663204ce13b: Already exists 
9e4017a20a07: Pull complete 
e57f3a8048c3: Pull complete 
d3a86e1b89cd: Pull complete 
Digest: sha256:f5bb83de6d4450d274ea9151531b41f5027a6a558c45fb839b64ca320b0863cf
Status: Downloaded newer image for mattermost/mm-ee-test:master
{"timestamp":"2022-08-09 08:15:49.199 Z","level":"info","msg":"Server is initializing...","caller":"app/server.go:276","go_version":"go1.18.1"}
```
```
mmm:~> docker run -it --platform linux/amd64 --rm mattermost/mm-ee-test:8660461
Unable to find image 'mattermost/mm-ee-test:8660461' locally
8660461: Pulling from mattermost/mm-ee-test
8663204ce13b: Already exists 
cb90c48655c8: Pull complete 
e66eac65df5c: Pull complete 
c8f606ffd2dc: Pull complete 
4f4fb700ef54: Pull complete 
Digest: sha256:6bb93a81103bf18b1578748d78d6a5e442041344efd9673a6918d381ace7034b
Status: Downloaded newer image for mattermost/mm-ee-test:8660461
{"timestamp":"2022-08-09 08:18:41.398 Z","level":"info","msg":"Server is initializing...","caller":"app/server.go:276","go_version":"go1.18.1"}
{"timestamp":"2022-08-09 08:18:41.398 Z","level":"info","msg":"Pinging SQL","caller":"sqlstore/store.go:228","database":"master"}
{"timestamp":"2022-08-09 08:18:41.399 Z","level":"error","msg":"Failed to ping DB","caller":"sqlstore/store.go:238","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
^Cmmm:~> docker run -it --platform linux/amd64 --rm mattermost/mm-ee-test:32a52f4
Error relocating /mattermost/bin/mattermost: fcntl64: symbol not found
mmm:~> 
```

**Comment by noxer** [comment]
Maybe this helps

**Comment by phoinixgrr** [comment]
It now fails with: 

```
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
[454:0809/113735.535991:ERROR:sandbox_linux.cc(377)] InitializeSandbox() called with multiple threads in process gpu-process.
[454:0809/113735.538610:ERROR:gpu_memory_buffer_support_x11.cc(44)] dri3 extension not supported.
Cypress could not verify that this server is running:
  > http://server:8065/
We are verifying this server because it has been configured as your baseUrl.
Cypress automatically waits until your server is accessible before running tests.
We will try connecting to it 3 more times...
We will try connecting to it 2 more times...
We will try connecting to it 1 more time...
Cypress failed to verify that your server is running.
Please start this server and then run Cypress again.
Successfully written: environment.json
Successfully updated the cycle with test environment data!
/usr/src/e2e/run_test_cycle.js:112
    const {stats, tests, spec} = result.runs[0];
                                            ^
TypeError: Cannot read properties of undefined (reading '0')
    at saveResult (/usr/src/e2e/run_test_cycle.js:112:45)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async runSpecFragment (/usr/src/e2e/run_test_cycle.js:253:5)
    at async runSpec (/usr/src/e2e/run_test_cycle.js:274:22)
    ```

**Comment by noxer** [comment]
Maybe this PR is just cursed? :ghost:

**Comment by phoinixgrr** [comment]
Maybe the author of the e2e tests, can shed some lights on this.

**Comment by Phrynobatrachus** [comment]
I'm looking again. At first glance that would appear to be a similar failure (server failed to start), but the image tag `mm-te-test:957b8d4` does actually start correctly on my machine, so it must be something else. :cry: 

Thank you for the investigation so far though.

**Comment by Phrynobatrachus** [comment]
#20794 build image has the same 

`Error relocating /mattermost/bin/mattermost: fcntl64: symbol not found`

issue, just noticed while trying to check the e2e pipeline against a different PR.

**Comment by agnivade** [comment]
@saturninoabril - Just wanted to check your pulse on if you think it's okay to merge this without e2e tests and check the results on master instead? We have been facing some random libc issues specifically related to e2e tests and this PR has nothing to do with the failures we are seeing. I somehow feel this PR has got stuck in some weird limbo state. Overall, this PR just improves the error logging for some JSON parsing. So relatively very low risk.

Let me know your thoughts.

**Comment by noxer** [comment]
> @noxer please could you rebase and solve conflicts just to be sure? Is there a companion ee branch for this?

Hmm, there are no merge conflicts, would a rebase change something significant? There is no ee branch for this.

**Review by saturninoabril — APPROVED** [review]
@phoinixgrr re: [e2e error](https://github.com/mattermost/mattermost-server/pull/20735#issuecomment-1209287921), it failed because server didn't start successfully.

@agnivade I can confirm that this is not related to E2E nor changes introduced from this PR. This is a build issue as also confirmed by @metanerd and @Phrynobatrachus.

Changes were tested locally and no obvious issue found on quick Cypress test.

LGTM and I won't block merging this PR. I would suggest for build issue to address separately. Feel free to merge anytime.

**Comment by phoinixgrr** [comment]
Thanks for letting us know. Raised  the following ticket to further investigate and mitigate this. 

https://mattermost.atlassian.net/browse/CLD-3929
https://mattermost.atlassian.net/browse/CLD-3916

**Comment by phoinixgrr** [comment]
@Phrynobatrachus @metanerd @saturninoabril

We have located the culprit, and the analysis is available @ https://mattermost.atlassian.net/browse/CLD-3929
We have raised https://mattermost.atlassian.net/browse/CLD-3942 to mitigate this.

**Comment by saturninoabril** [comment]
Thanks @phoinixgrr for the update!

**Pull request merged**
